presto-main/src/main/java/io/prestosql/util/LongLong2LongOpenCustomBigHashMap.java

```diff
@@ -108,6 +109,13 @@
      */
     protected long defRetValue;
 
+    /**
+     * The two-part value denoting unmapped keys (or null keys). These values may be passed back via the HashStrategy callback
+     * during equality checks, even though no keys with these values have been added.
+     */
+    protected final long nullKey1;
```

Comment:
1. Can we just assume that `(0, 0)` are special keys? Why do we have to make it selective?
2. In what way `HashStrategy` can be more clever regarding to special keys? Does that introduce additional branching for CPU?

Reply:
1. The fastutil code previously assumes (0,0) are actually special keys, which is a problem for us since all of our IDs are actually 0 index based, and overlaps in the numerical space. We won't be able to tell them apart from real entries.
2. The way we end up using this later like this: https://github.com/prestosql/presto/pull/6333/commits/c22eb4d4b6bd1434a2f811b88deefa173b1b06ae#diff-ed97f2619a0070f10f22c87f1cb1b5218c741a0e3eff94feb178d6eed616be24R103
We really don't want to expose internal non-allocated keys to our ID buffers. I can explain a bit more if you would like.

Reply:
The special null keys are not so much about performance as they are about having the HashStrategy providers be able to distinguish true null keys from keys that have already been submitted. (0, 0) actually is the default in fastutil, but the challenge here is that it overlaps with the key space we currently have -- and we need it to be more predictable to give the right answers. This is how we end up using this later:

https://github.com/prestosql/presto/pull/6333/commits/c22eb4d4b6bd1434a2f811b88deefa173b1b06ae#diff-ed97f2619a0070f10f22c87f1cb1b5218c741a0e3eff94feb178d6eed616be24R103